File: lib/services/zodiac-service.ts
// Servicio de zodíaco integrado en Next.js

interface ZodiacResponse {
  western_zodiac: string
  chinese_zodiac: string
}

export function getZodiacSign(dateOfBirth: string): ZodiacResponse {
  // Determinar el signo zodiacal occidental
  const [year, month, day] = dateOfBirth.split("-").map(Number)
  let westernZodiac = "Desconocido"

  if ((month === 3 && day >= 21) || (month === 4 && day <= 19)) westernZodiac = "Aries"
  else if ((month === 4 && day >= 20) || (month === 5 && day <= 20)) westernZodiac = "Tauro"
  else if ((month === 5 && day >= 21) || (month === 6 && day <= 20)) westernZodiac = "Géminis"
  else if ((month === 6 && day >= 21) || (month === 7 && day <= 22)) westernZodiac = "Cáncer"
  else if ((month === 7 && day >= 23) || (month === 8 && day <= 22)) westernZodiac = "Leo"
  else if ((month === 8 && day >= 23) || (month === 9 && day <= 22)) westernZodiac = "Virgo"
  else if ((month === 9 && day >= 23) || (month === 10 && day <= 22)) westernZodiac = "Libra"
  else if ((month === 10 && day >= 23) || (month === 11 && day <= 21)) westernZodiac = "Escorpio"
  else if ((month === 11 && day >= 22) || (month === 12 && day <= 21)) westernZodiac = "Sagitario"
  else if ((month === 12 && day >= 22) || (month === 1 && day <= 19)) westernZodiac = "Capricornio"
  else if ((month === 1 && day >= 20) || (month === 2 && day <= 18)) westernZodiac = "Acuario"
  else if ((month === 2 && day >= 19) || (month === 3 && day <= 20)) westernZodiac = "Piscis"

  // Determinar el signo zodiacal chino
  const animals = [
    "Rata",
    "Buey",
    "Tigre",
    "Conejo",
    "Dragón",
    "Serpiente",
    "Caballo",
    "Cabra",
    "Mono",
    "Gallo",
    "Perro",
    "Cerdo",
  ]
  const chineseZodiac = animals[(year - 4) % 12]

  return {
    western_zodiac: westernZodiac,
    chinese_zodiac: chineseZodiac,
  }
}

File: backend/zodiac_service.py
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
from base_service import create_service
from datetime import datetime

app, run = create_service("Zodiac", 5001)

class ZodiacRequest(BaseModel):
    date_of_birth: str

@app.post("/zodiac")
async def get_zodiac(request: ZodiacRequest):
    try:
        # Convertir la fecha de nacimiento a objeto datetime
        birth_date = datetime.strptime(request.date_of_birth, "%Y-%m-%d")
        
        # Determinar el signo zodiacal occidental
        month = birth_date.month
        day = birth_date.day
        
        western_zodiac = ""
        if (month == 3 and day >= 21) or (month == 4 and day <= 19):
            western_zodiac = "Aries"
        elif (month == 4 and day >= 20) or (month == 5 and day <= 20):
            western_zodiac = "Tauro"
        elif (month == 5 and day >= 21) or (month == 6 and day <= 20):
            western_zodiac = "Géminis"
        elif (month == 6 and day >= 21) or (month == 7 and day <= 22):
            western_zodiac = "Cáncer"
        elif (month == 7 and day >= 23) or (month == 8 and day <= 22):
            western_zodiac = "Leo"
        elif (month == 8 and day >= 23) or (month == 9 and day <= 22):
            western_zodiac = "Virgo"
        elif (month == 9 and day >= 23) or (month == 10 and day <= 22):
            western_zodiac = "Libra"
        elif (month == 10 and day >= 23) or (month == 11 and day <= 21):
            western_zodiac = "Escorpio"
        elif (month == 11 and day >= 22) or (month == 12 and day <= 21):
            western_zodiac = "Sagitario"
        elif (month == 12 and day >= 22) or (month == 1 and day <= 19):
            western_zodiac = "Capricornio"
        elif (month == 1 and day >= 20) or (month == 2 and day <= 18):
            western_zodiac = "Acuario"
        else:
            western_zodiac = "Piscis"
        
        # Determinar el signo zodiacal chino
        year = birth_date.year
        animals = ["Rata", "Buey", "Tigre", "Conejo", "Dragón", "Serpiente", 
                  "Caballo", "Cabra", "Mono", "Gallo", "Perro", "Cerdo"]
        chinese_zodiac = animals[(year - 4) % 12]
        
        return {
            "western_zodiac": western_zodiac,
            "chinese_zodiac": chinese_zodiac
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

if __name__ == "__main__":
    run()

